src/Autocomplete/src/Doctrine/EntityMetadata.php

```diff
@@ -41,22 +41,45 @@ public function isEmbeddedClassProperty(string $propertyName): bool
         return \array_key_exists($propertyNameParts[0], $this->metadata->embeddedClasses);
     }
 
-    public function getPropertyMetadata(string $propertyName): array
+    public function getFieldMetadata(string $propertyName): array
     {
         if (\array_key_exists($propertyName, $this->metadata->fieldMappings)) {
-            return $this->metadata->fieldMappings[$propertyName];
+            $metadata = $this->metadata->fieldMappings[$propertyName];
+
+            // In doctrine/orm:^3.0; $metadata will be a FieldMapping object
+            if (!\is_array($metadata)) {
+                return (array) $metadata;
+            }
+
+            return $metadata;
         }
 
+        throw new \InvalidArgumentException(sprintf('The "%s" field does not exist in the "%s" entity.', $propertyName, $this->metadata->getName()));
+    }
+
+    public function getAssociationMetadata(string $propertyName): array
```

Comment:
Nice idea splitting the 2 methods. Let's add `@internal` to these 2 methods at least. Then yes, leave the old method - we just won't call it anymore directly. And trigger a deprecation inside of it. We'll need to require the deprecations contract package, but that's no big deal.